cumulus/polkadot-omni-node/lib/src/common/runtime.rs

```diff
@@ -62,7 +101,109 @@ pub trait RuntimeResolver {
 pub struct DefaultRuntimeResolver;
 
 impl RuntimeResolver for DefaultRuntimeResolver {
-	fn runtime(&self, _chain_spec: &dyn ChainSpec) -> sc_cli::Result<Runtime> {
-		Ok(Runtime::Omni(BlockNumber::U32, Consensus::Aura(AuraConsensusId::Sr25519)))
+	fn runtime(&self, chain_spec: &dyn ChainSpec) -> sc_cli::Result<Runtime> {
+		let metadata_inspector = MetadataInspector::new(chain_spec)?;
+		let block_number = match metadata_inspector.block_number() {
+			Some(inner) => inner,
+			None => {
+				log::warn!(
+					r#"⚠️  There isn't a runtime type named `System`, corresponding to the `frame-system`
+                pallet (https://docs.rs/frame-system/latest/frame_system/). Please check Omni Node docs for runtime conventions:
+                https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/omni_node/index.html#runtime-conventions.
+                Note: We'll assume a block number size of `u32`."#
+				);
+				BlockNumber::U32
+			},
+		};
+
+		if !metadata_inspector.pallet_exists(DEFAULT_PARACHAIN_SYSTEM_PALLET_NAME) {
+			log::warn!(
+				r#"⚠️  The parachain system pallet (https://docs.rs/crate/cumulus-pallet-parachain-system/latest) is
+			   missing from the runtime’s metadata. Please check Omni Node docs for runtime conventions:
+			   https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/omni_node/index.html#runtime-conventions."#
+			);
+		}
+
+		Ok(Runtime::Omni(block_number, Consensus::Aura(AuraConsensusId::Sr25519)))
+	}
+}
+
+/// Inspect runtime metadata.
+pub struct MetadataInspector(subxt::Metadata);
```

Comment:
Does it need to be public?

Reply:
No strong opinion. I would hide it, and expose it when requested/needed.
If you want to keep it public, maybe a better doc would be nice - what exactly it allows to achieve.

Reply:
No strong opinions either. Making it public later if requested is better than making it public now and maybe needing to deprecate it (or move it somewhere else in the future and deprecate it here). Done this here: [0261c3b](https://github.com/paritytech/polkadot-sdk/pull/6450/commits/0261c3b427d39c1c884d381e3aa6308390cdae35). Thanks!